src/snowflake/snowpark/_internal/ast_utils.py

```diff
@@ -1325,3 +1328,45 @@ def build_expr_from_dict_str_str(
         t = ast_dict.add()
         t._1 = k
         t._2 = v
+
+
+def reset_snowpark_temp_ids(textproto: str) -> str:
+    id_lookup = {}
+    cnt = 0
+    for prefix, object_type, orig_id in re.findall(
+        rf"\"(SNOWPARK_TEMP_)({'|'.join([e.value for e in TempObjectType])})_(.*)\"",
+        textproto,
+    ):
+        if orig_id not in id_lookup:
+            cnt = cnt + 1
+            id_lookup[orig_id] = f"{cnt:010x}"
+        sub_id = id_lookup[orig_id]
+        textproto = textproto.replace(
+            f"{prefix}{object_type}_{orig_id}", f"{prefix}{object_type}_{sub_id}"
+        )
+    return textproto
+
+
+def base64_str_to_request(base64_str: str) -> proto.Request:
+    message = proto.Request()
+    message.ParseFromString(base64.b64decode(base64_str.strip()))
+    return message
+
+
+def base64_str_to_textproto(base64_str: str) -> str:
+    request = base64_str_to_request(base64_str)
+
+    # Force a fixed python version to avoid unnecessary diffs
+    request.client_language.python_language.version.major = 3
+    request.client_language.python_language.version.minor = 9
+    request.client_language.python_language.version.patch = 1
+    request.client_language.python_language.version.label = "final"
+
+    message = MessageToString(request)
+    message = reset_snowpark_temp_ids(message)
```

Comment:
Is it feasible to do this on the original proto message, instead of the text format? It seems suboptimal to do text processing when we have the original structured data at hand.